ArduCopter/events.cpp

```diff
@@ -4,152 +4,233 @@
+        action = FS_THR_DISABLED;
     } else {
-        if (control_mode == Mode::Number::AUTO && g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
-            // continue mission
-        } else if (flightmode->is_landing() &&
-                   battery.has_failsafed() &&
-                   battery.get_highest_failsafe_priority() <= FAILSAFE_LAND_PRIORITY) {
-            // continue landing or other high priority failsafes
-        } else {
-            if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_RTL) {
-                set_mode_RTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            } else if (g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
-                set_mode_RTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            } else if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_SMARTRTL_OR_RTL) {
-                set_mode_SmartRTL_or_RTL(MODE_REASON_RADIO_FAILSAFE);
-            } else if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_SMARTRTL_OR_LAND) {
-                set_mode_SmartRTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            } else { // g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_LAND
-                set_mode_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            }
-        }
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE");
     }
 
-    AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_RADIO, LogErrorCode::FAILSAFE_OCCURRED);
-
+    // Execute the specified action
+    switch(action) {
+        case FS_THR_DISABLED:
+            break;
+        case FS_THR_ENABLED_ALWAYS_RTL:
+        case FS_THR_ENABLED_CONTINUE_MISSION:
+            set_mode_RTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
+            break;
+        case FS_THR_ENABLED_ALWAYS_SMARTRTL_OR_RTL:
+            set_mode_SmartRTL_or_RTL(MODE_REASON_RADIO_FAILSAFE);
+            break;
+        case FS_THR_ENABLED_ALWAYS_SMARTRTL_OR_LAND:
+            set_mode_SmartRTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
+            break;
+        case FS_THR_ENABLED_ALWAYS_LAND:
+            set_mode_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
+            break;
+        default:
+            set_mode_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
+    }
 }
 
 // failsafe_off_event - respond to radio contact being regained
-// we must be in AUTO, LAND or RTL modes
-// or Stabilize or ACRO mode but with motors disarmed
 void Copter::failsafe_radio_off_event()
 {
     // no need to do anything except log the error as resolved
     // user can now override roll, pitch, yaw and throttle and even use flight mode switch to restore previous flight mode
     AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_RADIO, LogErrorCode::FAILSAFE_RESOLVED);
+    gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE CLEARED");
 }
 
-void Copter::handle_battery_failsafe(const char *type_str, const int8_t action)
+void Copter::handle_battery_failsafe(const char *type_str, int8_t action)
```

Comment:
Don't discard const here, you don't match the function signature anymore.

Reply:
Fixed